designs/warmreplicas.md

```diff
@@ -0,0 +1,151 @@
+Add support for warm replicas
+===================
+
+## Motivation
+Controllers reconcile all objects during startup / leader election failover to account for changes
+in the reconciliation logic. For certain sources, the time to serve the initial list can be
+significant in the order of minutes. This is problematic because by default controllers (and by
+extension watches) do not start until they have won leader election. This implies guaranteed
+downtime as even after leader election, the controller has to wait for the initial list to be served
+before it can start reconciling.
+
+## Proposal
```

Comment:
What if there's a lot of churn and the controller doesn't become leader maybe not at all or maybe after a few days?

The queue length will increase while there is nothing that takes items out of the queue.

I know the queue doesn't require significant memory to store an item but is there something we should be concerned about if the queue has eg millions of items (let's say we watch pods and we don't become leader for a month)

Reply:
Worst case it at some point gets oom killed, restarts, does everything again, I don't think this is likely to become an actual issue

Reply:
Yeah definitely not a big problem. Maybe just good to know

Reply:
> Doesn't the queue have deduplication built in? I thought that you couldn't re-add a key to the queue if the key already existed within the queue.

Yes there is de-duplication.

> queue length would always be equal to the number of those objects in the cluster

I don't think that deleted objects are removed from our queue. So if you have churn (i.e. any objects are deleted) the queue length won't be equal to the number of objects in the cluster